src/dataframe/dataframe.jl

```diff
@@ -1508,3 +1508,63 @@ end
 function permutecols!(df::DataFrame, p::AbstractVector{Symbol})
     permutecols!(df, index(df)[p])
 end
+
+##############################################################################
+##
+## Expand & Complete dataframes
+##
+##############################################################################
+
+
+function expanddf(df::AbstractDataFrame, indexcols::Array{Symbol,1})
+    colnames = names(df)
+    # Check to make sure the symbols in indexcols are in the df
+    for i in indexcols
+        if i ∉ colnames
+            throw(ArgumentError(":$i is not in dataframe"))
+        end
+    end
+    dummydf = similar(df[indexcols],0)
+
+    # Create a dictionary of symbol=>vector of unique values
+    uniqueVals = Dict{Symbol, Array{Any,1}}()
+    for col in indexcols
+        uniqueVals[col] = unique(df[col])
+    end
+
+    # Get a long vector of every possible combination
+    collected = collect(flatten(product(values(uniqueVals)...)))
+    dfsize = div(length(collected),length(indexcols))
+
+    # Re-format to dataframe
+    for i in 1:length(indexcols):length(collected)
+        push!(dummydf, collected[i:(i+length(indexcols)-1)])
+    end
+
+    return dummydf
+end
+
+function completedf(df::AbstractDataFrame, indexcols::Array{Symbol,1}; fill=missing::Any)
+    colnames = names(df)
+    # Check to make sure the symbols in indexcols are in the df
+    for i in indexcols
+        if i ∉ colnames
+            throw(ArgumentError(":$i is not in dataframe"))
+        end
+    end
+
+    # Expand the input df and left join
+    expanded = expanddf(df, indexcols)
+    expanded = join(expanded, df; on = indexcols, kind = :left)
+
+    # Replace missing values with the fill
```

Comment:
There is a tricky issue here that the original columns of `df` might have had contained missing values. I understand that they would be overwritten by `fill`. `join` actually allows you to distinguish case of `missing` originating from the `df` vs `missing` produced by `join` because you can temporarily generate the column that tells you if the source contained such a combination of values.

But maybe you wanted the implementation as it is now (or maybe we should give an option for the user to choose).

Reply:
dplyr seems to do that too:
> If you supply fill, these values will also replace existing explicit missing values in the data set.

But that doesn't seem like a great idea to me (you can do that manually if you want). Better distinguish these by default as you suggest.

Reply:
I like the idea of distinguishing the original missing values from ones created by this package.

Reply:
I added a kwarg to `complete` called `replaceallmissing=false`. By default (false) it will only replace the missing values created by the expansion using the `:source` column from the join. true will replace all missings.

Reply:
ok